src/concurrency/vector_clock.rs

```diff
@@ -50,21 +52,39 @@ const SMALL_VECTOR: usize = 4;
 /// so that diagnostics can report what code was responsible for an operation.
 #[derive(Clone, Copy, Debug)]
 pub struct VTimestamp {
-    time: u32,
+    // The lowest bit indicates read type, the rest is the time.
+    time_and_read_type: u32,
     pub span: Span,
 }
 
 impl VTimestamp {
-    pub const ZERO: VTimestamp = VTimestamp { time: 0, span: DUMMY_SP };
+    pub const ZERO: VTimestamp = VTimestamp { time_and_read_type: 0, span: DUMMY_SP };
+
+    fn time(&self) -> u32 {
+        self.time_and_read_type.shr(1)
+    }
+
+    fn set_time(&mut self, time: u32) {
+        self.time_and_read_type =
+            (self.time_and_read_type & 1) | time.checked_shl(1).expect("Vector clock overflow")
+    }
 
     pub fn span_data(&self) -> SpanData {
         self.span.data()
     }
+
+    pub fn read_type(&self) -> NaReadType {
+        if self.time_and_read_type & 1 == 0 { NaReadType::Read } else { NaReadType::Retag }
+    }
+
+    pub fn set_read_type(&mut self, read_type: NaReadType) {
+        self.time_and_read_type = (self.time_and_read_type & !1) | read_type as u32
+    }
```

Comment:
This seems to make some non-local assumptions about `Read` casting to 0 and `Retag` to 1? IMO it'd be better to keep the logic for int-to-NaReadType and the other direction next to each other to reduce the risk of them getting out-of-sync.